<launch>
  <!-- roslaunch arguments -->
  <arg name="show_rviz" default="false"/>
  <arg name="paused" default="false"/>
  <arg name="debug" default="false"/>
  <arg name="gui" default="false"/>
  <arg name="track_name" default="medium_track.world"/>
  <arg name="controller" default="false" />
  <arg name="dynamic_recon" default="false" />
  <arg name="with_auto_recovery" default="false" />
  <arg name="arrow" default="true" />
  <arg name="record" default="false" />
  <arg name="play" default="false" />
  <arg name="perspective" default="1" />
  <arg name="robot_pos" default="-x 1.6 -y 0 -z 0 -Y 1.57" />
  <arg name="camera_pos" default="-x 1.0 -y 0 -z 0 -Y 0" />
  <arg name="godeye_pos" default="-x 0 -y 0 -z 0 -Y 0" />

  <!-- We resume the logic in empty_world.launch, changing only the name of 
    the world to be launched -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find sim_world)/worlds/$(arg track_name)" />
    <arg name="paused" value="$(arg paused)" />
    <arg name="debug" value="$(arg debug)" />
    <arg name="gui" value="$(arg gui)" />    
    <arg name="use_sim_time" value="true" />
    <arg name="headless" value="false" />
  </include>

  <!-- urdf xml robot description loaded on the Parameter Server-->
  <param name="robot_description" command="$(find xacro)/xacro.py '$(find sim_environment)/urdf/wheel_robot.urdf.xacro'" />
  <param name="camera_description" command="$(find xacro)/xacro.py '$(find sim_environment)/urdf/surveillance_camera.urdf.xacro'" />
  <param name="godeye_description" command="$(find xacro)/xacro.py '$(find sim_environment)/urdf/godeye.urdf.xacro'" />
  
  <group if="$(eval perspective==1)">
    <param name="godeye_perspective_description" command="$(find xacro)/xacro.py '$(find sim_environment)/urdf/godeye_perspective.urdf.xacro'" />
  </group>
  <group if="$(eval perspective==2)">
    <param name="godeye_perspective_description" command="$(find xacro)/xacro.py '$(find sim_environment)/urdf/godeye_perspective2.urdf.xacro'" />
  </group>
  <group if="$(eval perspective==3)">
    <param name="godeye_perspective_description" command="$(find xacro)/xacro.py '$(find sim_environment)/urdf/godeye_perspective3.urdf.xacro'" />
  </group>

  <!-- push robot_description to factory and spawn robot in gazebo -->
  <node name="wheel_robot_spawn" pkg="gazebo_ros" type="spawn_model" output="screen" 
    args="-urdf $(arg robot_pos)  -param robot_description -model wheel_robot" />
  <node name="camera_spawn" pkg="gazebo_ros" type="spawn_model" output="screen" 
    args="-urdf $(arg camera_pos)  -param camera_description -model surveillance_camera" />
  <node name="godeye_spawn" pkg="gazebo_ros" type="spawn_model" output="screen" 
    args="-urdf $(arg godeye_pos)  -param godeye_description -model godeye_camera" />
  <group if="$(eval perspective==1)">
   <node name="godeye_perspective_spawn" pkg="gazebo_ros" type="spawn_model" output="screen" 
      args="-urdf -x 2.5 -y -3.5 -z 0.0   -param godeye_perspective_description -model godeye_camera_perspective" />  
  </group> 
  <group if="$(eval perspective==2)">
   <node name="godeye_perspective_spawn" pkg="gazebo_ros" type="spawn_model" output="screen" 
      args="-urdf -x 0 -y -3.7 -z 0.0   -param godeye_perspective_description -model godeye_camera_perspective" />  
  </group>
  <group if="$(eval perspective==3)">
   <node name="godeye_perspective_spawn" pkg="gazebo_ros" type="spawn_model" output="screen" 
      args="-urdf -x 7.5 -y -10.5 -z 0.0   -param godeye_perspective_description -model godeye_camera_perspective" />  
  </group>
 
  <!-- robot visualization in Rviz -->
  <group if="$(arg show_rviz)">
    <node name="rviz" pkg="rviz" type="rviz" output="screen" args="-d $(find sim_environment)/launch/wheel_robot.rviz"/> 
  </group> 

  <!-- load joint controller configurations from YAML file to parameter server -->
  <rosparam file="$(find sim_environment)/config/joint_position_control.yaml" command="load" />
  <!-- load the controllers -->
  <node name="controller_manager" pkg="controller_manager" type="spawner" respawn="false"
    output="screen" args="rear_left_wheel_velocity_controller rear_right_wheel_velocity_controller
                                        front_left_wheel_velocity_controller front_right_wheel_velocity_controller
                                        front_left_hinge_position_controller front_right_hinge_position_controller
                                        joint_state_controller"/>

  <!-- publish all the frames to TF -->
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher">
    <param name="publish_frequency" value="50"/> <!-- Hz -->
  </node>
  <!--<node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" ></node>-->
  
  <!--<node name="image_view_sur_camera" pkg="image_view" type="image_view" args="image:=/surveillance_camera/image_raw" />-->
  <group if="$(eval play==false)">
    <node name="image_view_godeye_camera" pkg="image_view" type="image_view" args="image:=/godeye_camera/image_raw" />
   </group>
  <group unless="$(eval play==false)">
    <node name="image_view_godeye_camera_w_champ" pkg="image_view" type="image_view" args="image:=/godeye_camera_w_champ/image_raw" />
  </group>

  <!--<node name="rqt_gui" pkg="rqt_gui" type="rqt_gui"/>-->
  <group if="$(eval arrow==true)">
    <node name="image_view" pkg="image_view" type="image_view" args="image:=/front_camera_w_arrow/image_raw" />
  </group>
  <group if="$(eval arrow==false)">
    <node name="image_view" pkg="image_view" type="image_view" args="image:=/front_camera/image_raw" />
  </group>

  <node name="image_view_godeye_camera_perspective" pkg="image_view" type="image_view" args="image:=/godeye_camera_perspective/image_raw" />
  
  <!-- use dynamic reconfigure course_out_surveillance-->
  <group if="$(eval dynamic_recon==true)">
    <!-- dynamic reconfigure SERVER for max speed coeff node -->
      <node pkg="sim_environment" type="dynamic_recon_server.py" name="dynamic_recon_server_node" output="screen" />
    <!-- dynamic reconfigure CLIENT for max speed coeff node -->
      <node pkg="sim_environment" type="course_out_surveillance.py" name="course_out_surveillance" output="screen" />
  </group>

  <!-- servo node -->
    <node pkg="sim_environment" type="servo_commands.py" name="servo_commands" output="screen" />

  <!-- field information manamegent node -->
  <!-- timing surveillance node-->
    <node pkg="sim_environment" type="judge_communication_d.py" name="judge_communication_d" output="screen" />  
  <!-- timing surveillance node-->
    <node pkg="sim_environment" type="timing_surveillance.py" name="timing_surveillance" output="screen" />
  <!-- courseout recovery node-->
    <node pkg="sim_environment" type="jugemu.py" name="jugemu" output="screen" />  
  <group if="$(eval with_auto_recovery==true)">
    <rosparam file="$(find sim_environment)/config/respown_point.yaml" command="load" />
    <!-- for large field -->
    <rosparam file="$(find sim_environment)/config/respown_point_large.yaml" command="load" /> 
    <node pkg="sim_environment" type="detect_stack.py" name="detect_stack" output="screen" /> 
    <node pkg="sim_environment" type="jugemu_new.py" name="jugemu_new" output="screen" />
  </group>

  <group unless="$(eval play==false)">
    <node pkg="sim_environment" type="position_player.py" name="champion_replay" args="$(arg play)" />
  </group>

  <group if="$(eval record==true)">
    <node pkg="sim_environment" type="position_recording.py" name="record_position"/>
  </group>

  <!-- without gui(with keyboard con) -->
  <group if="$(eval arrow==false)">
    <group unless="$(eval play==false)">
      <node pkg="sim_environment" type="window_management2.py" name="window_management" output="screen" args="-play" />
    </group>
    <group if="$(eval play==false)">
      <node pkg="sim_environment" type="window_management2.py" name="window_management" output="screen" />
    </group>
  </group>
  <group if="$(eval arrow==true)">
    <group unless="$(eval play==false)">
      <node pkg="sim_environment" type="window_management2.py" name="window_management" output="screen" args="-play -arrow" />
    </group>
    <group if="$(eval play==false)">
      <node pkg="sim_environment" type="window_management2.py" name="window_management" output="screen" args="-arrow"/>
    </group>
  </group>
  <group if="$(eval arrow==true)">
    <node pkg="sim_environment" type="command_Superimposition.py" name="add_arrow"/>
  </group>
  <group if="$(eval controller==true)">
      <node pkg="utility" type="keyboard_con_pygame2.py" name="keyboard_con_pygame2" output="screen" />
  </group>

</launch>
